Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp

```diff
@@ -300,13 +311,26 @@ WebCore::DisplayRefreshMonitor& ThreadedCompositor::displayRefreshMonitor() cons
 void ThreadedCompositor::frameComplete()
 {
     ASSERT(m_compositingRunLoop->isCurrent());
+    displayUpdateFired();
     sceneUpdateFinished();
 }
 
 void ThreadedCompositor::targetRefreshRateDidChange(unsigned rate)
 {
     ASSERT(RunLoop::isMain());
-    m_displayRefreshMonitor->setTargetRefreshRate(rate);
+
+    if (!rate)
+        rate = c_defaultRefreshRate;
+    m_display.displayUpdate = { 0, rate / 1000 };
+}
+
+void ThreadedCompositor::displayUpdateFired()
+{
+    m_display.displayUpdate = m_display.displayUpdate.nextUpdate();
+
+    WebProcess::singleton().eventDispatcher().notifyScrollingTreesDisplayDidRefresh(m_display.displayID);
```

Comment:
Why did you remove the `Client::displayDidRefresh`? `EventDispatcher::notifyScrollingTreesDisplayDidRefresh` iterates all scrolling trees to notify that a display has been refreshed. Every tree checks if its displayID matches the given one. This is required when there's a global display notifying about a refresh, but in our case we have one display refresh monitor with its own displayID for every threaded compositor. So, we know what tree we wat to notify without having to iterate all trees.

Reply:
This is called from the composition thread, so `LayerTreeHost::displayDidRefresh()` ends up loading the Page's ScrollingCoordinator and then its ScrollingTree from a non-main thread.

Using EventDispatcher avoids that, and it locks over the scrolling trees that are registered and unregistered on the main thread, so no threading problems.

Reply:
I see, in that case I would add:

```
void EventDispatcher::notifyWebPageScrollingTreeDisplayDidRefresh(WebPage& page, PlatformDisplayID displayID);
```

And call it from `LayerTreeHost::displayDidRefresh`. This avoids the tree iteration, fixes the threading issues and doesn't use WebProcess API from Shared.

Reply:
I assume you would want to short-cut to the ScrollingTree in the new EventDispatcher method, through the WebPage. But that still operates on that WebPage on a non-main thread, so threading problems would remain.

Calling `ThreadedScrollingTree::displayDidRefresh()` on every scrolling tree isn't ideal, and it can be improved, but I don't see it as critical.

Reply:
It would be just getting the identifier that is the key of the trees map. Anyway, it's not critical since in most of the cases there's only one page per process. But still I prefer to call the dispatcher from `LayerTreeHost::displayDidRefresh`, which is in WebProcess layer too. At some point we should move the threaded compositor code from Shared to WebProcess and stop pretending it's shared, though.

Reply:
Yes, I'm adding back `LayerTreeHost::displayDidRefresh`.